Tracing: Trace system overhaul

tleksell-pe opened the pull request:
Adds generic trace hook macros for inserting trace hook for
function entry, exit, blocking, and object initialization
with a variable number of arguments (objects, return values,
etc.). Utilizing macro concatenation these macros produce
trace hooks of a similar format to the old trace system
with `SYS_TRACING_OBJ_FUNC(k_thread, switched_in)` being
turned into `sys_trace_k_thread_switched_in()` by the
preprocessor. Although these macros still rely on the manual
definition of each unique trace hook in tracing.h, the benefit
of not directly calling those is that we can enable/disable
trace hooks based on object type (`k_thread`, `k_sem`, etc.)
through the preprocessor while providing the ability of adding
type specific runtime trace processing similar to
`SYS_TRACING_OBJ_INIT`.

Adds tracing hooks to most kernel objects to enable tracing of
core behavior/state and functionality that affects application
scheduling. This improves on the previous tracing hooks which
only covered a fraction of the important events within Zephyr.

Adds syscall tracing by injecting syscall entry/exit tracing
hooks as `__syscall` functions are parsed by the `gen_syscalls.py`
script.

Note 1: Since this tracing system overhaul currently co-exists
with the previous system, all new trace hooks and macros use a
`sys_port_trace` or `SYS_PORT_TRACING` notation to avoid conflicts.
This notation with an extra port/PORT infix should be removed
if this system is adopted over the previous system.

Note 2: To enable tracing of return values many new trace hooks
required the decomposition of return statements. Although these
changes should be safe it is important that the respective code
owners verify that there are no unintended side-effects.

Note 3: Tracing of calls like `k_msleep` can cause problems
with User Mode threads since these will potentially invoke
trace code that is outside their domain. If possible these
should be migrated to have their own `__syscall` wrapper similar
to `k_sleep`.

Note 4: Several static inline functions defined in kernel.h
were changed into declarations with definitions in their
respective kernel object source file. This was done to
facilitate tracing and avoid including the entire tracing
system in `kernel.h`.

nashif commented:
@tleksell-pe thanks for the PR.

I just noticed you mention `SYS_TRACING_OBJ_INIT`. This implementation predates the main tracing feature we have supporting systemview and ctf. The intent of OBJECT_TRACING was to maintain the list of objects in a way that would allow tracking usage of objects in gdb and does not provide the level of tracing capabilities and integration with tools possible with current tracing implementation and the goodness coming with this PR. My thought is that once we have this PR integrated and have all objects covered we should be able to deprecate the old OBJECT_TRACING and move completely to the new design.

tleksell-pe commented:
@nashif Thanks for the info, if we would go ahead and deprecate the old OBJECT_TRACING system we might also consider moving the new tracing macros from the object_tracing_common.h file into the tracing header or tracing subsystem for the sake of clarity. This should also make it easier to override the trace macros in situation where a tracing system wants to do specific preprocessor injection or additional filtering based on their requirements.

tleksell-pe committed:
Tracing: Trace hook support macros

Adds generic trace hook macros for inserting trace hook for
function entry, exit, blocking, and object initialization
with a variable number of arguments (objects, return values,
etc.). Utilizing macro concatenation these macros produce
trace hooks of a similar format to the old trace system
with SYS_TRACING_OBJ_FUNC(k_thread, switched_in) being
turned into sys_trace_k_thread_switched_in() by the
preprocessor. Although these macros still rely on the manual
definition of each unique trace hook in tracing.h, the benefit
of not directly calling those is that we can enable/disable
trace hooks based on object type (k_thread, k_sem, etc.)
through the preprocessor while providing the ability of adding
type specific runtime trace processing similar to
SYS_TRACING_OBJ_INIT.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Semaphore tracing

Add semahphore tracing using the new trace macros.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Semaphore tracing documentation

Add default semaphore trace hooks and documentation into
tracing.h.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Mutex tracing hooks

Add mutex trace hooks, default mutex trace hooks, and trace hook
documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Conditional variable tracing

Add conditional variable tracing hooks, default tracing hooks,
and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Thread tracing

Add thread tracing hooks, default hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Queue tracing

Add Queue tracing hooks, default hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: FIFO Tracing

Add FIFO tracing hook, default hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: LIFO Tracing

Add LIFO tracing hooks, default hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Memory Stack tracing

Add memory stack tracing, defaul trace hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Message Queue tracing

Add Message Queue tracing, default hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Mailbox tracing

Add Mailbox tracing, default tracing hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Pipe tracing

Add Pipe tracing, default trace hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Memory Heap tracing

Add Memory heap tracing, default trace hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Memory Slab tracing

Add memory slab tracing, default trace hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Timer tracing

Add Timer tracing, default tracing hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Poll API and Work Poll tracing

Add Poll API and Work Poll tracing, default hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

tleksell-pe committed:
Tracing: Work tracing

Add Work tracing, default tracing hooks, and documentation.

Signed-off-by: Torbjörn Leksell <torbjorn.leksell@percepio.com>

nashif commented on subsys/tracing/Kconfig:
```diff
@@ -1,225 +1,331 @@
-endif
-
-source "subsys/tracing/sysview/Kconfig"
+# Debug configuration options
```
looks like you have Windows line termination here..

```
file  subsys/tracing/Kconfig
subsys/tracing/Kconfig: ASCII text, with CRLF line terminators
```

tleksell-pe replied:
Good find, this has now been resolved.

nashif reviewed (changes requested):
In general all looks good with a few comments related to how we can move forward:
- do not mix tracing with existing object tracing, they are not related. existing object tracing code will be removed once we have coverage for all objects.
- Need to start removing old tracing macros and just leave the new ones where applicable.
- remove place holder comments
- Remove code instead of just commenting it out.
- as an indicator that all works as expected, we need to make the macros first  work with current tracing backends, at least the CTF backend. I can do that on top of you PR.

Thanks.

nashif commented:
added #34371 draft for removing object tracing

nashif commented:
@tleksell-pe  I took the liberty and rebased this PR on top of latest master.

tleksell-pe commented on include/kernel.h:
```diff
@@ -1928,10 +1913,11 @@ static inline int z_impl_k_queue_is_empty(struct k_queue *queue)
 __syscall void *k_queue_peek_head(struct k_queue *queue);
 
-static inline void *z_impl_k_queue_peek_head(struct k_queue *queue)
+/*static inline void *z_impl_k_queue_peek_head(struct k_queue *queue)
```
Fixed, I'll make sure to update similar occurrences if found.

tleksell-pe commented on include/kernel.h:
```diff
@@ -1944,10 +1930,11 @@ static inline void *z_impl_k_queue_peek_head(struct k_queue *queue)
 __syscall void *k_queue_peek_tail(struct k_queue *queue);
 
-static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
+/*static inline void *z_impl_k_queue_peek_tail(struct k_queue *queue)
```
Fixed, I'll make sure to update similar occurrences if found.

tleksell-pe commented on include/kernel.h:
```diff
@@ -19,6 +19,7 @@
 #include <limits.h>
 #include <stdbool.h>
 #include <toolchain.h>
+#include "debug/object_tracing_common.h"
```
It is currently required for the LIFO/FIFO macros in kernel.h, given a migration of the macros from "object_tracing_common.h" it would look better if the new/existing trace header was included if tracing is enabled.

tleksell-pe replied:
There is now a tracing_macros.h file in the tracing directory which is included in this spot instead.

tleksell-pe commented on kernel/sem.c:
```diff
@@ -66,17 +66,23 @@ int z_impl_k_sem_init(struct k_sem *sem, unsigned int initial_count,
+	/* [TZ-TRACE]: New trace hook */
+	SYS_PORT_TRACING_OBJ_FUNC(k_sem, init, sem, 0);
+
 	sys_trace_semaphore_init(sem);
```
Sure thing, I've removed all except four, the ISR and idle hooks since they are called from assembly code.

tleksell-pe commented on include/debug/object_tracing_common.h:
```diff
@@ -20,6 +20,20 @@
+/**
+ * @note	Extended tracing macro blanks
+ */
+#define SYS_PORT_TRACING_FUNC(type, func, ...) do { } while (false)
```
Absolutely, it would be possible to put them in tracing.h but it is fairly stuffed at the moment. We could go for a separate header file in the same directory as tracing.h if that is fine with you guys.

nashif replied:
yes, this should be fine.

tleksell-pe replied:
These macros have now been migrated into a tracing_macros.h file in the tracing directory.

nordic-krch commented:
It may be possible to add a generic wrapper to every api call to avoid adding enter/exit macros inside a function (which becomes problematic for maintenance with functions with multiple return locations). I've sketched a proposal: https://github.com/zephyrproject-rtos/zephyr/commit/7f4a03bd73ac48f8b75e22a31a291710134f8306
It is locally redefining a function name to redirect the call to a wrapping macro.

However, there are few things to consider:
- no userspace handling in the proposal
- it adds wrapping to every kernel api call so increases code size
- it is using GCC feature (https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html#Statement-Exprs) but it is already used in the PR

nashif commented:
closing this one, the bulk of those changes was already merged
